### ISSUE TITLE
GPU training stalls >60 s compiling native artifact for XGBoost predictor (nv0)

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#497]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(24534500727251).epochs(1).log(all).stop(0.8).int(1).save(&bundle).run(&model, &data)` on backend `nv0` remained active after 60 seconds. The model includes `recipe.model().xgbst().log().norm(batch).iq(4).nl .xgbst().log().norm(batch).iq(1).m`. The earliest failing operation is the compilation of the native artifact for the NVIDIA backend: `recipe.train().run()` → `compile_model` → `NativeModelIr::from_graph` → `emit` → `compile_native_artifact` (NVIDIA branch, PTX/ptxas) for the predictor primitive produced by `fit_xgboost` (see `recipe.rs` around line 7356‑7385). The data loader (`data/temporal/series_sqlite.sqlite`) and subsequent model blocks are downstream of this compile step and do not affect the earliest failure.

No existing issue in the repository covers an XGBoost predictor compilation stall on the NVIDIA `nv0` backend. Issue #346 addresses LightGBM predictor stalls, and issue #372 addresses KNN stalls, but the implementation path for XGBoost (`fit_xgboost` → `boosted_predictor` → NVIDIA native compilation) is distinct.

Therefore this packet represents a new performance defect requiring a new issue.

**Expected runtime**: For a single‑epoch training on the modest SQLite temporal dataset with two XGBoost blocks, the GPU compilation and execution should complete in a few seconds. A 60‑second stall indicates a material slowdown caused by the native artifact compilation step.

**Conclusion**: New issue needed for XGBoost predictor native‑artifact compilation stall on backend `nv0`.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `ed8550390fa6e670`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"Public `recipe.train().optimizer(adamw).lr(0.001).seed(24534500727251).epochs(1).log(all).stop(0.8).int(1).save(&bundle).run(&model, &data)` on backend `nv0` remained active after 60 seconds. The model includes `recipe.model().xgbst().log().norm(batch).iq(4).nl .xgbst().log().norm(batch).iq(1).m`. The earliest failing operation is the compilation of the native artifact for the NVIDIA backend: `recipe.train().run()` → `compile_model` → `NativeModelIr::from_graph` → `emit` → `compile_native_artifact` (NVIDIA branch, PTX/ptxas) for the predictor primitive produced by `fit_xgboost` (see `recipe.rs` around line 7356‑7385). The data loader (`data/temporal/series_sqlite.sqlite`) and subsequent model blocks are downstream of this compile step and do not affect the earliest failure.\n\nNo existing issue in the repository covers an XGBoost predictor compilation stall on the NVIDIA `nv0` backend. Issue #346 addresses LightGBM predictor stalls, and issue #372 addresses KNN stalls, but the implementation path for XGBoost (`fit_xgboost` → `boosted_predictor` → NVIDIA native compilation) is distinct.\n\nTherefore this packet represents a new performance defect requiring a new issue.\n\n**Expected runtime**: For a single‑epoch training on the modest SQLite temporal dataset with two XGBoost blocks, the GPU compilation and execution should complete in a few seconds. A 60‑second stall indicates a material slowdown caused by the native artifact compilation step.\n\n**Conclusion**: New issue needed for XGBoost predictor native‑artifact compilation stall on backend `nv0`.","issue":0,"rationale":"Tracing the code path shows the earliest public failure is the native artifact compilation for the NVIDIA backend (`compile_model` → `compile_native_artifact`) triggered by the XGBoost estimator (`fit_xgboost`). This path is not represented by any existing issue (LightGBM stalls are issue #346, KNN stalls are #372). The measured 60 s runtime far exceeds the expected sub‑second compile time for this small dataset, confirming a new performance regression specific to XGBoost on nv0.","title":"GPU training stalls >60 s compiling native artifact for XGBoost predictor (nv0)","verdict":"new"}
```

## Deterministic failure packet

kind=performance
id=ed8550390fa6e670
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:14689 next:14690 composition:24534500727251
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-40-14689/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/series_sqlite.sqlite")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.xgbst().log().norm(batch).iq(4).nl
		.xgbst().log().norm(batch).iq(1).m
		.loss(rmse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(24534500727251)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.int(1)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(71298284147480).epochs(1).log(all).stop(0.0).fp(8).save(&bundle).run(&model, &data)` on `nv0` crossed the 60 s threshold. The model is `recipe.model().xgbst().log().qi(4).0 .cbst().tan().norm(batch).qi(5).0 .loss(mae)`. The loader `recipe.data("data/temporal/chronological_splits").target("target").split(0.8)` feeds a small temporal dataset. The earliest public failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (NVIDIA `ptx`/`ptxas` branch) for the predictor primitive produced by `.xgbst()` (`fit_xgboost` in `recipe.rs`). The `.cbst()` block and quantization settings are downstream and do not change the earliest cause. A small dataset with default boosted-predictor iterations should compile and complete in a few seconds on nv0; the 60 s active state is materially slower, dominated by the native-artifact compilation stall for the XGBoost predictor, exactly the cause reported in issue #497.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `24d4a611631d7f98`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict": "comment", "issue": 497, "title": "Performance: one-epoch nv0 stall on xgbst/cbst temporal split matches existing XGBoost predictor compilation stall", "body": "Public `recipe.train().optimizer(adamw).lr(0.001).seed(71298284147480).epochs(1).log(all).stop(0.0).fp(8).save(&bundle).run(&model, &data)` on `nv0` crossed the 60 s threshold. The model is `recipe.model().xgbst().log().qi(4).0 .cbst().tan().norm(batch).qi(5).0 .loss(mae)`. The loader `recipe.data(\"data/temporal/chronological_splits\").target(\"target\").split(0.8)` feeds a small temporal dataset. The earliest public failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (NVIDIA `ptx`/`ptxas` branch) for the predictor primitive produced by `.xgbst()` (`fit_xgboost` in `recipe.rs`). The `.cbst()` block and quantization settings are downstream and do not change the earliest cause. A small dataset with default boosted-predictor iterations should compile and complete in a few seconds on nv0; the 60 s active state is materially slower, dominated by the native-artifact compilation stall for the XGBoost predictor, exactly the cause reported in issue #497.", "rationale": "The packet reaches the same earliest operation (`compile_native_artifact` for `.xgbst()` on nv0) through the same implementation path (`fit_xgboost` -> `boosted_predictor` -> NVIDIA native compilation) described in issue #497. The differences (`.cbst()` second block, `.qi(4).0` vs `.iq(4).nl`, `.fp(8)`, `data/temporal/chronological_splits`, `seed(71298284147480)`) are irrelevant configuration changes that occur after or parallel to the earliest failing step and do not establish a different public defect. Therefore the packet is additional evidence for the existing issue."}
```

## Deterministic failure packet

kind=performance
id=24d4a611631d7f98
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:13960 next:13961 composition:71298284147480
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-489-13960/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.xgbst().log().qi(4).0
		.cbst().tan().norm(batch).qi(5).0
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(71298284147480)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.fp(8)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
The public `recipe.train().run()` remains active past 60 seconds on `nv0` during `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` for the predictor graph containing `.xgbst()`. The temporal loader and preceding `.forest(2)` fitting complete before this native-artifact compilation stall, matching the existing XGBoost predictor issue.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `f5a4751e6fc9e69b`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 497,
  "title": "GPU training stalls >60 s compiling native artifact for XGBoost predictor (nv0)",
  "body": "The public `recipe.train().run()` remains active past 60 seconds on `nv0` during `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` for the predictor graph containing `.xgbst()`. The temporal loader and preceding `.forest(2)` fitting complete before this native-artifact compilation stall, matching the existing XGBoost predictor issue.",
  "rationale": "The earliest failing public operation is native artifact compilation before epoch execution, reached through the NVIDIA `ptx`/`ptxas` path for the `.xgbst()` predictor. Source inspection shows `lower_block` invokes estimator fitting while building the graph, then `compile_model` emits and compiles the resulting predictor artifact. The workload is 12,000 rows with 1 feature, 2 forest trees, 8 boosting iterations, and depth 5. Forest and XGBoost fitting perform roughly 50 sorted tree levels over 12,000 samples, with low tens of millions of comparisons and modest memory traffic, so a defensible expected runtime is sub-second to a few seconds on nv0. The measured 60 seconds is materially unjustified and dominated by native artifact compilation. Issue #497 represents the same nv0 XGBoost predictor compilation cause; `.forest(2)`, temporal input, quantization, and loss do not establish a different earliest defect."
}
```

## Deterministic failure packet

kind=performance
id=f5a4751e6fc9e69b
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:23706 next:23707 composition:53717658789294
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-8162-23706/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target");
	let model = recipe.model()
		.forest(2).elu().qi(2).k
		.xgbst().tan().qi(5).k
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(53717658789294)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(64)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
The public training run remains active after 60 seconds on nv0. The earliest failure is train preparation while lowering the first `.xgbst().exp().qi(3).k` block: `lower_estimator` calls `fit_xgboost`, which performs 8 boosting iterations with depth-5 trees and repeatedly sorts all 493,537 training rows for split candidates before native execution begins. The workload is roughly 248 row-subset sorts, followed by kmeans fitting and predictor compilation; a defensible expected range is approximately 10 to 30 seconds on this host, so the measured 60 seconds is materially slow. The dominant gap is the unbounded host-side XGBoost tree fitting and its large generated predictor, matching the existing nv0 XGBoost predictor stall.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `6ae77900826ef2de`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 497,
  "title": "NVIDIA nv0 training stalls beyond 60s on XGBoost predictor",
  "body": "The public training run remains active after 60 seconds on nv0. The earliest failure is train preparation while lowering the first `.xgbst().exp().qi(3).k` block: `lower_estimator` calls `fit_xgboost`, which performs 8 boosting iterations with depth-5 trees and repeatedly sorts all 493,537 training rows for split candidates before native execution begins. The workload is roughly 248 row-subset sorts, followed by kmeans fitting and predictor compilation; a defensible expected range is approximately 10 to 30 seconds on this host, so the measured 60 seconds is materially slow. The dominant gap is the unbounded host-side XGBoost tree fitting and its large generated predictor, matching the existing nv0 XGBoost predictor stall.",
  "rationale": "Issue search identified #497 as the existing nv0 XGBoost predictor performance issue, and its full issue record and comments were inspected. Source inspection confirms that `recipe.train().run()` calls `prepare`, then `compile` and `lower_estimator`; `fit_xgboost` executes before `compile_model` or any epoch arithmetic. The exact dataset contains 508,537 loaded rows, 493,537 source training rows, and one feature. XGBoost uses the configured 8 boosting iterations and depth 5, with up to 31 split nodes per tree, while the downstream `.kmeans(2)` and quantization do not change the earliest cause. The 60-second measurement is additional evidence for the same XGBoost predictor path rather than a distinct defect."
}
```

## Deterministic failure packet

kind=performance
id=6ae77900826ef2de
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:22824 next:22825 composition:80307365974776
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-6733-22824/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target");
	let model = recipe.model()
		.xgbst().exp().qi(3).k
		.kmeans(2).gelu().norm(batch).qi(3).k.l
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(80307365974776)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0
